## Secant Activity 

To estimate the values through root-finding with the Secant Method, I'll use the code below (from class).

In [2]:
def secant_method(f, x0, x1, max_iter = 100, tol = 1e-6):
    """
    Secant method for finding a root of the equation f(x) = 0.

    Parameters:
    f        : Function for which we want to find the root.
    x0, x1   : Initial guesses for the root.
    tol      : Tolerance for stopping criterion.
    max_iter : Maximum number of iterations.

    Returns:
    Approximate root of the function.
    """
    for i in range(max_iter):
        # Calculate the value of the function at the guesses
        f_x0 = f(x0)
        f_x1 = f(x1)
        
        # Secant method formula to find the next approximation
        x2 = x1 - f_x1 * (x1 - x0) / (f_x1 - f_x0)
        
        # Check for convergence
        if abs(x2 - x1) < tol:
            print(f"Converged after {i+1} iterations.")
            return x2
        
        # Update guesses for the next iteration
        x0, x1 = x1, x2
    
    print("Warning: Maximum iterations reached without convergence.")
    return x2

### Problem 1

Find the root of $f(x) = x^2 − 2$, to estimate $\sqrt{2}$ starting with $x_0 = 1$ and $x_1 = 2$. Do a maximum number of 10 iterations.

In [6]:
def f(x):
    return x**2 - 2 

# Initial guesses
x0 = 1.0
x1 = 2.0

# Max number of iterations
max_iter = 10

# Call secant_method to find the root
root = secant_method(f, x0, x1, max_iter)
print(f"Approximate root: {root}")

Converged after 6 iterations.
Approximate root: 1.4142135623730954


So using the Secant method, $\sqrt{2} \approx 1.4142$. We can trust this estimation, as the method converged after 6 iterations, indicating that the estimation is very close to the actual value at this point. If the method did not converge, or converged on 10, it would be a worse estimation.

### Problem 2

Find the root of $f(x) = e^x − 2$, to estimate $\ln{(2)}$ starting with $x_0 = 0$ and $x_1 = 3$. Do a maximum number of 10 iterations.

In [ ]:
import math

def f(x):
    return math.exp(x) - 2

# Initial guesses
x0 = 0.0
x1 = 3.0

# Max number of iterations
max_iter = 10

# Call secant_method to find the root
root = secant_method(f, x0, x1, max_iter)
print(f"Approximate root: {root}")

Converged after 8 iterations.
Approximate root: 0.6931471805596721


So using the Secant method, $\ln{(2)} \approx 0.6931$. We can trust this result, for the same reasons discussed above.